src/app/main-page/presentation/toolbar-action/toolbar-action.component.ts

```diff
@@ -128,14 +124,64 @@ export class ToolbarActionComponent implements OnChanges {
         this.graphType.emit(this.graphSettings.type);
     }
 
-    setMapType(mapType: MapType) {
-        this.mapType.emit(mapType);
+    setMapData(mapVariant: MapVariant): void {
+        // please note: MapType.BLACK_AND_WHITE is temporarily deactivated
+        const isMap =
+            mapVariant ===
+            Constants.DEFAULT_MAP_VARIANTS[
+                MapType.MAPNIK
+            ]; /*|| mapVariant === Constants.DEFAULT_MAP_VARIANTS[MapType.BLACK_AND_WHITE]*/
+        const isShapeFileOnMap =
+            mapVariant ===
+            Constants.DEFAULT_MAP_VARIANTS[MapType.SHAPE_FILE_ON_MAP];
+        const callback = () => {
+            this.setMapType(mapVariant);
+        };
+        let newMapVariant: MapVariant = mapVariant;
+
+        if (isMap) {
+            const { mapLayer: newMapLayer } = mapVariant;
+            this.setMapTypeSelected(newMapLayer!, callback);
+            return;
+        }
+
+        if (isShapeFileOnMap) {
+            const { lastMapTypeSelected } = this.graphSettings;
+            newMapVariant = {
+                ...mapVariant,
+                mapLayer: lastMapTypeSelected,
+            };
+        }
+
+        this.setMapType(newMapVariant);
+    }
+
+    setMapTypeSelected(
+        newMapLayer: MapType.MAPNIK /*|MapType.BLACK_AND_WHITE*/,
+        callback,
+    ): void {
+        // please note: MapType.BLACK_AND_WHITE is temporarily deactivated
+        const { lastMapTypeSelected } = this.graphSettings;
+        const differentMapSelected = newMapLayer !== lastMapTypeSelected;
+
+        if (!differentMapSelected) {
+            callback();
+            return;
+        }
+
+        this.mapTypeSelected.emit(newMapLayer);
+        callback();
+    }
```

Comment:
same here: tried to stick to early returns structure and destructuring as it helps, in case funcs get more complex later on. but either way is fine for me.